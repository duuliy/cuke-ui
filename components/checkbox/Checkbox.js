import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import cls from "classnames";

export default class Checkbox extends PureComponent {
  static defaultProps = {
    prefixCls: "cuke-checkbox",
    defaultChecked: false,
    indeterminate: false,
    checked: false,
    isButton: false,
    onChange: () => {},
    size: "default"
  };
  static propTypes = {
    prefixCls: PropTypes.string.isRequired,
    onChange: PropTypes.func,
    disabled: PropTypes.bool,
    checked: PropTypes.bool,
    defaultChecked: PropTypes.bool,
    indeterminate: PropTypes.bool,
    size: PropTypes.oneOf(["small", "default", "large"])
  };
  state = {
    checked: this.props.checked || this.props.defaultChecked
  };
  _onChange = e => {
    this.setState(({ checked }) => {
      return {
        checked: !checked
      };
    });
    this.props.onChange(e);
  };

  // eslint-disable-next-line
  UNSAFE_componentWillReceiveProps({ checked }) {
    if (checked !== this.props.checked) {
      this.setState({
        checked
      });
    }
  }

  render() {
    const {
      className,
      prefixCls,
      disabled,
      children,
      value,
      isButton,
      size,
      onChange, //eslint-disable-line
      indeterminate,
      ...attr
    } = this.props;
    const { checked } = this.state;
    return (
      <label
        className={cls(`${prefixCls}-wrapper`, {
          [`${prefixCls}-button-wrapper`]: isButton,
          [`${prefixCls}-checked`]: checked,
          [`${prefixCls}-disabled`]: disabled,
          [`${prefixCls}-indeterminate`]: checked && indeterminate,
          [`${prefixCls}-${size}`]: isButton
        })}
        {...attr}
      >
        <span
          className={cls(prefixCls, className, {
            [`${prefixCls}-button`]: isButton
          })}
        >
          <input
            type="checkbox"
            value={value}
            checked={checked}
            className={cls(`${prefixCls}-input`)}
            onChange={this._onChange}
            disabled={disabled}
          />
          <span className={cls(`${prefixCls}-inner`)} />
        </span>
        <span>{children}</span>
      </label>
    );
  }
}
