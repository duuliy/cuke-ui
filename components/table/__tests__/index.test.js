import assert from "power-assert";
import React from "react";
import { render, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Table from "../index";

const dataSource = [
  {
    name: "黄瓜",
    count: 1,
    id: 1
  },
  {
    name: "西瓜",
    count: 344,
    id: 2
  },
  {
    name: "香蕉",
    count: 199,
    id: 3
  }
];

const columns = [
  {
    title: "ID",
    dataIndex: "id",
    key: "id"
  },
  {
    title: "名字",
    dataIndex: "name",
    key: "name"
  },
  {
    title: "数量",
    dataIndex: "count",
    key: "count"
  },
  {
    title: "操作",
    dataIndex: "setting",
    key: "setting",
    render: (value, data, index) => <Button>删除</Button>
  }
];

import Button from "../../button";
import Empty from "../../empty";
import Pagination from "../../pagination";
import Spin from "../../spin";
import Checkbox from "../../checkbox";

describe("<Table/>", () => {
  it("should render Table", () => {
    const wrapper = render(<Table columns={columns} dataSource={dataSource} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should find cuke-table class name", () => {
    const wrapper = shallow(
      <Table columns={columns} dataSource={dataSource} />
    );
    assert(wrapper.find(".cuke-table").length === 1);
    assert(wrapper.find(".cuke-table-origin-table").length === 1);
    assert(wrapper.find(".cuke-table-pagination").length === 1);
    assert(wrapper.find(".cuke-table-tbody").length === 1);
  });
  it("should render empty table", () => {
    const wrapper = shallow(<Table columns={columns} dataSource={[]} />);
    assert(wrapper.find(Empty).length === 1);
    assert(wrapper.find(Pagination).length === 0);
    assert(wrapper.find(".cuke-table-tbody").length === 0);
  });

  it("should find custom render", () => {
    const wrapper = render(<Table columns={columns} dataSource={dataSource} />);
    expect(wrapper.text()).toContain("删除");
  });

  it("should can not render Pagination when pagination is false", () => {
    const wrapper = shallow(
      <Table columns={columns} dataSource={dataSource} pagination={false} />
    );
    assert(wrapper.find(Pagination).length === 0);
  });

  it("should render Spin when is loading", () => {
    const wrapper = shallow(
      <Table columns={columns} dataSource={dataSource} loading />
    );
    assert(wrapper.find(Spin).length === 1);
  });

  it("should render border", () => {
    const wrapper = shallow(
      <Table columns={columns} dataSource={dataSource} bordered />
    );
    assert(wrapper.find(".cuke-table-bordered").length === 1);
  });

  it("should render row selection", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Table
        columns={columns}
        dataSource={dataSource}
        rowSelection={{
          onChange,
          getCheckboxProps(record) {
            console.log("record: ", record);
          }
        }}
      />
    );
    assert(wrapper.find(Checkbox).length === 4);
    wrapper
      .find(Checkbox)
      .at(0)
      .simulate("change", {
        target: {
          checked: true
        }
      });
    expect(onChange).toHaveBeenCalled();
  });
  it("should cannot checked is disabled rowSelection when select all", () => {
    let _selectedRowKeys;
    const wrapper = shallow(
      <Table
        columns={columns}
        dataSource={dataSource}
        rowSelection={{
          onChange(selectedRowKeys) {
            _selectedRowKeys = selectedRowKeys;
          },
          getCheckboxProps(record) {
            return {
              disabled: record.name === "黄瓜1"
            };
          }
        }}
        pagination={{
          pageIndex: 1,
          pageSize: 5
        }}
        dataSource={new Array(30).fill().map((_, i) => ({
          name: `黄瓜${i + 1}`,
          count: i + 1,
          id: i + 1,
          key: i
        }))}
      />
    );
    wrapper
      .find(Checkbox)
      .at(0)
      .simulate("change", {
        target: {
          checked: true
        }
      });
    expect(
      wrapper
        .find(Checkbox)
        .at(1)
        .props().checked
    ).toBe(false);
    expect(_selectedRowKeys).toHaveLength(4);
  });

  it("should can not render table header", () => {
    const wrapper = shallow(
      <Table columns={columns} dataSource={dataSource} showHeader={false} />
    );
    assert(wrapper.find(".cuke-table-thead").length === 0);
  });

  it("should render stripe tr", () => {
    const wrapper = shallow(
      <Table columns={columns} dataSource={dataSource} stripe />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    assert(wrapper.find(".cuke-table-stripe").length === 1);
  });
});
