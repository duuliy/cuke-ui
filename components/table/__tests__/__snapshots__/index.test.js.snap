// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table/> should render Table 1`] = `
<div
  class="cuke-table"
  columns="[object Object],[object Object],[object Object],[object Object]"
>
  <table
    class="cuke-table-origin-table"
  >
    <thead
      class="cuke-table-thead"
    >
      <tr>
        <th>
          ID
        </th>
        <th>
          名字
        </th>
        <th>
          数量
        </th>
        <th>
          操作
        </th>
      </tr>
    </thead>
    <tbody
      class="cuke-table-tbody"
    >
      <tr
        class=""
      >
        <td>
          1
        </td>
        <td>
          黄瓜
        </td>
        <td>
          1
        </td>
        <td>
          <button
            class="cuke-button cuke-button-default cuke-button-normal"
            type="button"
          >
            <span>
              删除
            </span>
          </button>
        </td>
      </tr>
      <tr
        class=""
      >
        <td>
          2
        </td>
        <td>
          西瓜
        </td>
        <td>
          344
        </td>
        <td>
          <button
            class="cuke-button cuke-button-default cuke-button-normal"
            type="button"
          >
            <span>
              删除
            </span>
          </button>
        </td>
      </tr>
      <tr
        class=""
      >
        <td>
          3
        </td>
        <td>
          香蕉
        </td>
        <td>
          199
        </td>
        <td>
          <button
            class="cuke-button cuke-button-default cuke-button-normal"
            type="button"
          >
            <span>
              删除
            </span>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="cuke-table-pagination"
  >
    <ul
      class="cuke-pagination cuke-pagination-default"
      current="1"
      pageindex="1"
    >
      <li
        class="cuke-pagination-show-total"
      />
      <li
        class="cuke-pagination-item cuke-pagination-prev cuke-pagination-item-disabled"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <polyline
            points="15 18 9 12 15 6"
          />
        </svg>
      </li>
      <li
        class="cuke-pagination-item cuke-pagination-item-selected"
      >
        1
      </li>
      <li
        class="cuke-pagination-item cuke-pagination-next cuke-pagination-item-disabled"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <polyline
            points="9 18 15 12 9 6"
          />
        </svg>
      </li>
    </ul>
  </div>
</div>
`;

exports[`<Table/> should render stripe tr 1`] = `
<div
  className="cuke-table"
  columns={
    Array [
      Object {
        "dataIndex": "id",
        "key": "id",
        "title": "ID",
      },
      Object {
        "dataIndex": "name",
        "key": "name",
        "title": "名字",
      },
      Object {
        "dataIndex": "count",
        "key": "count",
        "title": "数量",
      },
      Object {
        "dataIndex": "setting",
        "key": "setting",
        "render": [Function],
        "title": "操作",
      },
    ]
  }
>
  <Spin
    indicator={<LoadingIcon />}
    prefixCls="cuke-spin"
    size="large"
    spinning={false}
    tip=""
  >
    <table
      className="cuke-table-origin-table"
    >
      <thead
        className="cuke-table-thead"
      >
        <tr>
          <th
            key="thead-0"
          >
            ID
          </th>
          <th
            key="thead-1"
          >
            名字
          </th>
          <th
            key="thead-2"
          >
            数量
          </th>
          <th
            key="thead-3"
          >
            操作
          </th>
        </tr>
      </thead>
      <tbody
        className="cuke-table-tbody"
      >
        <tr
          className=""
          key="id"
        >
          <td
            key="td-id"
          >
            1
          </td>
          <td
            key="td-name"
          >
            黄瓜
          </td>
          <td
            key="td-count"
          >
            1
          </td>
          <td
            key="td-setting"
          >
            <Button
              block={false}
              circle={false}
              dashed={false}
              disabled={false}
              hollow={false}
              href=""
              htmlType="button"
              loading={false}
              plain={false}
              prefixCls="cuke-button"
              size="default"
              type="default"
            >
              删除
            </Button>
          </td>
        </tr>
        <tr
          className="cuke-table-stripe"
          key="name"
        >
          <td
            key="td-id"
          >
            2
          </td>
          <td
            key="td-name"
          >
            西瓜
          </td>
          <td
            key="td-count"
          >
            344
          </td>
          <td
            key="td-setting"
          >
            <Button
              block={false}
              circle={false}
              dashed={false}
              disabled={false}
              hollow={false}
              href=""
              htmlType="button"
              loading={false}
              plain={false}
              prefixCls="cuke-button"
              size="default"
              type="default"
            >
              删除
            </Button>
          </td>
        </tr>
        <tr
          className=""
          key="count"
        >
          <td
            key="td-id"
          >
            3
          </td>
          <td
            key="td-name"
          >
            香蕉
          </td>
          <td
            key="td-count"
          >
            199
          </td>
          <td
            key="td-setting"
          >
            <Button
              block={false}
              circle={false}
              dashed={false}
              disabled={false}
              hollow={false}
              href=""
              htmlType="button"
              loading={false}
              plain={false}
              prefixCls="cuke-button"
              size="default"
              type="default"
            >
              删除
            </Button>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      className="cuke-table-pagination"
    >
      <Pagination
        current={1}
        locale={
          Object {
            "nextText": <ArrowRightIcon />,
            "prevText": <ArrowLeftIcon />,
          }
        }
        onChange={[Function]}
        onPageSizeChange={[Function]}
        pageIndex={1}
        pageSize={10}
        pageSizeOptions={
          Array [
            10,
            20,
            30,
            40,
          ]
        }
        prefixCls="cuke-pagination"
        separator="/"
        showQuickJumper={false}
        showSizeChanger={false}
        showTotal={[Function]}
        simple={false}
        size="default"
        total={3}
      />
    </div>
  </Spin>
</div>
`;
